src/abstraction/ak.rs

```diff
@@ -0,0 +1,184 @@
+                },
+            }))
+            .with_unique(PublicIdUnion::Ecc(Box::new(Default::default()))),
+    }
+    .build()
+}
+
+/// This loads an Attestation Key previously generated under the Endorsement hierarchy
+pub fn load_ak(
+    context: &mut Context,
+    parent: KeyHandle,
+    ak_auth_value: Option<&Auth>,
+    private: Private,
+    public: TPM2B_PUBLIC,
+) -> Result<KeyHandle> {
+    let session = match context.start_auth_session(
+        None,
+        None,
+        None,
+        SessionType::Policy,
+        Cipher::aes_128_cfb(),
+        HashingAlgorithm::Sha256,
+    )? {
+        Some(ses) => ses,
+        None => return Err(Error::local_error(WrapperErrorKind::WrongValueFromTpm)),
+    };
+    let session_attr = TpmaSessionBuilder::new()
+        .with_flag(TPMA_SESSION_DECRYPT)
+        .with_flag(TPMA_SESSION_ENCRYPT)
+        .build();
+    context.tr_sess_set_attributes(session, session_attr)?;
+
+    let _ = context.execute_with_nullauth_session(|ctx| {
+        ctx.policy_secret(
+            session,
+            AuthHandle::Endorsement,
+            Default::default(),
+            Default::default(),
+            Default::default(),
+            None,
+        )
+    })?;
+
+    let key_handle =
+        context.execute_with_session(Some(session), |ctx| ctx.load(parent, private, public))?;
+
+    context.flush_context(session.handle().into())?;
+
+    if let Some(ak_auth_value) = ak_auth_value {
+        context.tr_set_auth(key_handle.into(), ak_auth_value)?;
+    }
+
+    Ok(key_handle)
+}
+
+/// This creates an Attestation Key in the Endorsement hierarchy
+pub fn create_ak(
+    context: &mut Context,
+    parent: KeyHandle,
+    hash_alg: HashingAlgorithm,
+    sign_alg: SignatureScheme,
+    ak_auth_value: Option<&Auth>,
+) -> Result<CreateKeyResult> {
+    let key_alg = sign_alg.get_key_alg();
+
+    let ak_pub = create_ak_public(key_alg, hash_alg, sign_alg)?;
+
+    let session = match context.start_auth_session(
+        None,
+        None,
+        None,
+        SessionType::Policy,
+        Cipher::aes_128_cfb(),
+        HashingAlgorithm::Sha256,
+    )? {
+        Some(ses) => ses,
+        None => return Err(Error::local_error(WrapperErrorKind::WrongValueFromTpm)),
+    };
+    let session_attr = TpmaSessionBuilder::new()
+        .with_flag(TPMA_SESSION_DECRYPT)
+        .with_flag(TPMA_SESSION_ENCRYPT)
+        .build();
+    context.tr_sess_set_attributes(session, session_attr)?;
+
+    let _ = context.execute_with_nullauth_session(|ctx| {
+        ctx.policy_secret(
+            session,
+            AuthHandle::Endorsement,
+            Default::default(),
+            Default::default(),
+            Default::default(),
+            None,
+        )
+    })?;
+
+    let loaded_key = context.execute_with_session(Some(session), |ctx| {
+        ctx.create_key(parent, &ak_pub, ak_auth_value, None, None, None)
+    })?;
+
+    context.flush_context(session.handle().into())?;
```

Comment:
Just realised - you'll need to make it so that this is called whenever a previous operation fails, before `?`. Same for `load_ak`, of course

Reply:
Fixed

Reply:
I just added an `execute_with_temporary_object`, since this is a case I've seen more often.